hpy/debug/src/debug_handles.c

```diff
@@ -75,7 +85,7 @@ void DHPy_invalid_handle(HPyContext *dctx, DHPy dh)
 {
     HPyDebugInfo *info = get_info(dctx);
     HPyContext *uctx = info->uctx;
-    assert(as_DebugHandle(dh)->is_closed);
+    assert(as_DebugHandle(dh)->is_closed || as_DebugHandle(dh)->is_context_constant);
     if (HPy_IsNull(info->uh_on_invalid_handle)) {
         // default behavior: print an error and abort
         HPy_FatalError(uctx, "Invalid usage of already closed handle");
```

Comment:
The error messages could be more specific (also for other scenarios) and we could pass some ID of the error to the user provided hook, but I am leaving that for followup work